staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types_jsonschema.go

```diff
@@ -196,6 +196,13 @@ type ValidationRule struct {
 	// If unset, the message is "failed rule: {Rule}".
 	// e.g. "must be a URL with the host matching spec.host"
 	Message string
+	// MessageExpression represents a CEL expression that can be evaluated when validation fails, whereas Message
+	// is just a static string. MessageExpression must statically evaluate to a string to be accepted.
+	// If both Message and MessageExpression are present on a Rule, then MessageExpression will be used if validation
+	// fails. If MessageExpression results in a runtime error, then Message will be used instead if non-empty. If
+	// Message is empty, then the default error message is used instead. In either case, the CEL error that MessageExpression
+	// failed with will be logged.
```

Comment:
Can we simplify the conditional logic in the text here? I found it a bit twisty to read.

Reply:
Would you prefer it as a list, maybe something like:

```markdown
If MessageExpression results in a runtime error, then it will be logged, and the fallback message will instead use:

* Message (if non-empty)
* The default error message (if Message is empty)
```

I left "If both Message and MessageExpression are present on a Rule, then MessageExpression will be used if validation fails" out of the edit since I thought the hard part to understand was with the runtime error resolution part, but let me know if that's not the case.

Reply:
"If MessageExpression results in a runtime error, the validation failure message is produced as-if the MessageExpression field were unset" ?

Reply:
Gotcha, sounds good to me. Fixed.